### PR TITLE
Started the first few code blocks for review, and i put them in an in…

### PR DESCRIPTION
**Pull request by ChrisForti**
…stall file.

- Was able to use `test -f plus/file/path` to search for a file, It seams to work fine?
- Removed all the apt-get commands, and all the commands worked so I incorporated that into my new install script.
- Have not tested on a clean dockerless OS.
- Also got rid of the ca-certificates, YOU were right that was two commands my `which ca-certificates curl` was searching for just curl.lol
# what is ca-certificates?
- Short - It is a digital certificate that is used to verify the identity of 3rd parties, and encrypt data between you and said 3rd party.

- Long - A certificate authority (ca), is a trusted entity (like Comodo) that issues digital documents that can be used like a digital passport. When we go to a website (like google.com), Google will send you their certificate. Our browsers do a little magic with the ca-certificate, and either verify or reject the certificate provided by Google.

### CODE REVIEW

File: dockers-install.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+if (which curl)
+then 
+  echo "curl already installed"
+else
+  echo "Installing curl"
+  sudo apt update
+  sudo apt install curl
+fi
+
+if (test -f /etc/apt/keyrings/docker.asc)
```
Comment (jasondoze) [inline dockers-install.sh]: What about the one we use in class [ -f /.../.../... ] ?

**Review by jasondoze — CHANGES_REQUESTED** [review]
Instead of me reviewing this code, lets try combining what you have that you think works into one script, on one branch. Delete the other branches after that and continue to commit to the same branch until its approved. Then run those conditionals one at a time in the terminal on a freshly imaged pi that is completely wiped. No configuration. 

Always test your code first with a clean image, otherwise you wont know what does or doesnt work. After you run a single conditional, navigate to where it was supposed to do something and see that it did it. Then run the same cond  again and see if it works i.e. it shouldnt keep running the else statement. 

Put a # comment above each cond explaining in detail what you are doing and why you are doing it that way you get good at this. Downloading deps, adding repos and keys are very very common in configuration so lets do it right.

**Comment by ChrisForti** [comment]
- I am pretty sure that is what I did, it's a new branch new script file???
- `which` does not search for a file it searches for a program, and keyring is a file that should be in docker.asc The test command also does not return  a path. It standard outputs the the file location, which can be used imo to produce cleaner code.
```
if (echo " keyring exists in $(test -f /usr/bin/keyring)")
else
  echo "installing the thing"
  apt Install thing
fi
```
- I do not push any code unless it's been tested. In this case tested on a clean OS. What I meant was once I have tested my code in the terminal. Some of the configuration steps are executed. So technically it's not clean anymore. So when I run the script, after my code is working, by testing in the terminal.the else statements are not firing off. Because have already created said file, or whatever  I am testing.

**Review by stephenlauck — APPROVED** [review]

**Comment by stephenlauck** [comment]
Yeah to test if a file exists you can do a `test -f FILEPATH` or you could `stat FILEPATH`, both will set the exit code correctly if it's there or not. I also like to use `test -s FILEPATH` which checks if the file is there AND not empty or zero.
> 
> ```
> if (echo " keyring exists in $(test -f /usr/bin/keyring)")
> else
>   echo "installing the thing"
>   apt Install thing
> fi
```

**Review by ChrisForti — COMMENTED** [review]
**Error message**
- Option /etc/apt/keyrings/docker.asc: 
   - Configuration item specification must have an =**<val>**

**Comment by ChrisForti** [comment]
<val>

File: dockers-install.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  sudo apt install curl
+fi
+
+# Adding, and configuration of gpg keyring
+if (test -f /usr/bin/keyring/docker.asc)
+then
+  echo "gpg keyring exists, and configured"
+else
+  echo "configuring gpg keyring"
+  sudo apt install -m 0755 -d /etc/apt/keyrings curl-fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc chmod a+r /etc/apt/keyrings/docker.asc
```
Comment (jasondoze) [inline dockers-install.sh]: @ChrisForti Drop that chunk of code we were looking at the other night in after the else

Comment (stephenlauck) [reply dockers-install.sh]: if you look at the docker install on their docs maybe follow that and just add tests for each step:

[c](https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository)

Each line would be an if statement with a test.

**Comment by ChrisForti** [comment]
- **Current error msg**
```E:**** Unsupported file /etc/apt/keyrings given on commandline
configuring keyring
chmod: cannot access '/etc/apt/keyrings/docker.asc': No such file or directory
/usr/bin/gpg
gpg keyring exists, and configured```

**Comment by ChrisForti** [comment]
- Entire output
```chris@docker2:~ $ bash test-docker.sh 
Hit:1 http://deb.debian.org/debian bookworm InRelease
Hit:2 http://deb.debian.org/debian-security bookworm-security InRelease
Hit:3 http://deb.debian.org/debian bookworm-updates InRelease    
Hit:4 http://archive.raspberrypi.com/debian bookworm InRelease   
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
16 packages can be upgraded. Run 'apt list --upgradable' to see them.
Installing ca-certificates
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
ca-certificates is already the newest version (20230311).
0 upgraded, 0 newly installed, 0 to remove and 16 not upgraded.
/usr/bin/curl
curl already installed
Creating keyring
Reading package lists... Done
E: Unsupported file /etc/apt/keyrings given on commandline
configuring keyring
chmod: cannot access '/etc/apt/keyrings/docker.asc': No such file or directory
/usr/bin/gpg
gpg keyring exists, and configured```

File: dockers-install.sh [hunk 3]
```diff
@@ -19,11 +19,28 @@ else
...
 # Adding, and configuration of gpg keyring
-if (test -f /usr/bin/keyring/docker.asc)
+if (which gpg)
 then
   echo "gpg keyring exists, and configured"
 else
   echo "configuring gpg keyring"
-  sudo apt install -m 0755 -d /etc/apt/keyrings curl-fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc chmod a+r /etc/apt/keyrings/docker.asc
+  sudo apt install /etc/apt/keyrings
+  sudo curl-fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc
```
Comment (stephenlauck) [inline dockers-install.sh]: missing a space here between `curl -fsSL` between calling curl and it's flags.

**Review by stephenlauck — CHANGES_REQUESTED** [review]

**Comment by jasondoze** [comment]
@stephenlauck u ever use this to create dirs?
`sudo install -m 0755 -d /etc/apt/keyrings`

**Comment by ChrisForti** [comment]
Run it without the apt? @stephenlauck 
- `sudo install etc/apt/keyrings`
- or are you asking if I have ran it like that?
- I did not know that was creating a directory?
- That would explain the path, I thought it was a file. 
- which technically is what a directory is..... huh

**Comment by ChrisForti** [comment]
- OHH I see what your saying the docs are
- `sudo install -m 0755 -d /etc/apt/keyrings`, and I added the apt. lol

**Comment by ChrisForti** [comment]
- Made the changes, and all tested out except for the `chmod` command
**current error**
- chmod: cannot access '/etc/apt/keyrings/docker.asc': No such file or directory

File: dockers-install.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
+# Update
+sudo apt update
+
+# Installing ca-certificates
+if (which ca-certificates)
+then
+  echo "ca-certificates already installed"
+else
+  echo "Installing ca-certificates"
+  sudo apt install ca-certificates
```
Comment (stephenlauck) [inline dockers-install.sh]: add a `-y` so it's `sudo apt install -y ca-certificates`

File: dockers-install.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+fi
+
+# Installing curl
+if (which curl)
+then
+  echo "curl already installed"
+else
+  echo "Installing curl"
+  sudo apt install curl
```
Comment (stephenlauck) [inline dockers-install.sh]: Add `-y`

File: dockers-install.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+fi
+
+# Creating a keyrings directory
+if (test -f /etc/bin/keyrings)
```
Comment (stephenlauck) [inline dockers-install.sh]: This is a directory not a file so use `test -d /etc/bin/keyrings`

File: dockers-install.sh [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+then
+  echo "creating keyrings"
+else
+  echo "Creating keyrings"
+  sudo install -m 0755 -d /etc/apt/keyrings
+fi
+
+# Adding gpg keyring, and configuring
+if (which gpg)
```
Comment (stephenlauck) [inline dockers-install.sh:32]: The check here would be checking the saved file not if gpg is installed like this:

`if (gpg /etc/apt/keyrings/docker.asc)`

Also you have 2 commands here, one that writes the key and another that changes it's mode using `chmod`. Maybe break out the chmod command and try to write a test for it?

**Review by stephenlauck — CHANGES_REQUESTED** [review]

File: dockers-install.sh [hunk 8]
```diff
@@ -37,3 +37,11 @@ else
   sudo curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc; sudo chmod a+r /etc/apt/keyrings/docker.asc
 fi
 
+if (test -f /etc/bin/keyrings/docker.asc)
```
Comment (stephenlauck) [inline dockers-install.sh]: you should check out how to test for the correct mode here, something like:

`stat -c "%a" /etc/bin/keyrings/docker.asc`

Comment (stephenlauck) [reply dockers-install.sh]: And then make sure you get the number you want, based on the permissions for `a+r`.

Comment (ChrisForti) [reply dockers-install.sh]: - I will work on that tonight, I am Imaging a fresh SD card now. Thank you sir!

Comment (ChrisForti) [reply dockers-install.sh]: - Also found this earlier today, and it is custom to PI OS lite 64, which what I am using.
```curl -fsSL https://download.docker.com/linux/debian/gpg | sudo -H gpg --dearmor -o /etc/apt/keyrings/docker.gpg```

File: dockers-install.sh (hunk 7 above)
Comment (ChrisForti) [reply dockers-install.sh:32]: I did have it like that, and put the command on the end of the install block to see if it worked that way.
```if (test -f /etc/bin/keyrings/docker.asc)
then
  echo " gpg keyring already configured"
else
  echo "configuring gpg keyrings"
  sudo chmod a+r /etc/apt/keyrings/docker.asc
fi```

File: dockers-install.sh (hunk 8 above)
Comment (ChrisForti) [reply dockers-install.sh]: - The issue is the .asc or .gpg which is what the `chmod` command is doing with the command above It is testing out now
- https://forums.docker.com/t/save-docker-repository-signing-key-correctly-in-asc-format/128373

Comment (ChrisForti) [reply dockers-install.sh]: - I Think I got it, but then the terminal freezes
- Current readout
```chris@docker:~ $ bash test-docker.sh 
Hit:1 http://deb.debian.org/debian bookworm InRelease
Hit:2 http://deb.debian.org/debian-security bookworm-security InRelease
Hit:3 http://deb.debian.org/debian bookworm-updates InRelease    
Hit:4 http://archive.raspberrypi.com/debian bookworm InRelease   
Get:5 http://deb.debian.org/debian bookworm/main Sources [9,489 kB]
Get:6 http://deb.debian.org/debian bookworm/non-free Sources [77.9 kB]                                    
Get:7 http://deb.debian.org/debian bookworm/non-free-firmware Sources [6,168 B]                           
Get:8 http://deb.debian.org/debian bookworm/contrib Sources [51.3 kB]                                     
Get:9 http://deb.debian.org/debian-security bookworm-security/contrib Sources [856 B]                     
Get:10 http://deb.debian.org/debian-security bookworm-security/main Sources [90.8 kB]                     
Get:11 http://deb.debian.org/debian-security bookworm-security/non-free-firmware Sources [796 B]          
Get:12 http://deb.debian.org/debian bookworm-updates/non-free-firmware Sources [2,076 B]                  
Get:13 http://deb.debian.org/debian bookworm-updates/contrib Sources [776 B]                              
Get:14 http://deb.debian.org/debian bookworm-updates/non-free Sources [1,544 B]                           
Get:15 http://deb.debian.org/debian bookworm-updates/main Sources [17.4 kB]                               
Fetched 9,738 kB in 11s (866 kB/s)                                                                        
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
3 packages can be upgraded. Run 'apt list --upgradable' to see them.
Installing ca-certificates
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
ca-certificates is already the newest version (20230311).
0 upgraded, 0 newly installed, 0 to remove and 3 not upgraded.
/usr/bin/curl
curl already installed
Creating keyrings
/usr/bin/gpg
gpg keyring exists, and configured
configuring gpg keyrings
File '/etc/apt/keyrings/docker.gpg' exists. Overwrite? (y/N) y










I can just type here ??????????
```

Comment (ChrisForti) [reply dockers-install.sh:40]: After i press y, no prompt, and I can type in the blank area.

Comment (ChrisForti) [reply dockers-install.sh:40]: I am going to push the code up, so you can see what I did.

**Comment by ChrisForti** [comment]
**Sweet!! I got it to work. I got up early and retested everything down the line, with a fresh mind**
- I was not testing right, last night
- I learned that I can not jump around while in testing mode
- I was changing too much w/o getting the right exit code, so I was not even sure what was, and was not working
- I am going to create a new branch with the working script, and push it.
- I now have to right a test for this
 ```
echo \
  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian \
  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
  sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
sudo apt update
```
- which means I need to research a few of these commands mainly `tee`

**Comment by ChrisForti** [comment]
**This shit is hard, but I guess if it were easy everyone would be doing it!**
